vm/systemSmartContracts/staking_test.go

```diff
@@ -1 +1,448 @@
+
+	retCode := stakingSmartContract.Execute(arguments)
+	assert.Equal(t, vmcommon.UserError, retCode)
+}
+
+func TestStakingSC_ExecuteStakeWrongUnmarshalDataShouldErr(t *testing.T) {
+	t.Parallel()
+
+	stakeValue := big.NewInt(0)
+	eei := &mock.SystemEIStub{}
+	eei.GetStorageCalled = func(key []byte) []byte {
+		return []byte("data")
+	}
+	stakingSmartContract, _ := NewStakingSmartContract(stakeValue, eei)
+	arguments := CreateVmContractCallInput()
+	arguments.Function = "stake"
+
+	retCode := stakingSmartContract.Execute(arguments)
+	assert.Equal(t, vmcommon.UserError, retCode)
+}
+
+func TestStakingSC_ExecuteStakeRegistrationDataStakedShouldErr(t *testing.T) {
+	t.Parallel()
+
+	stakeValue := big.NewInt(0)
+	eei := &mock.SystemEIStub{}
+	eei.GetStorageCalled = func(key []byte) []byte {
+		registrationDataMarshalized, _ := json.Marshal(&stakingData{Staked: true})
+		return registrationDataMarshalized
+	}
+	stakingSmartContract, _ := NewStakingSmartContract(stakeValue, eei)
+	arguments := CreateVmContractCallInput()
+	arguments.Function = "stake"
+
+	retCode := stakingSmartContract.Execute(arguments)
+	assert.Equal(t, vmcommon.UserError, retCode)
+}
+
+func TestStakingSC_ExecuteStakeNotEnoughArgsShouldErr(t *testing.T) {
+	t.Parallel()
+
+	stakeValue := big.NewInt(0)
+	eei := &mock.SystemEIStub{}
+	eei.GetStorageCalled = func(key []byte) []byte {
+		registrationDataMarshalized, _ := json.Marshal(&stakingData{})
+		return registrationDataMarshalized
+	}
+	stakingSmartContract, _ := NewStakingSmartContract(stakeValue, eei)
+	arguments := CreateVmContractCallInput()
+	arguments.Function = "stake"
+
+	retCode := stakingSmartContract.Execute(arguments)
+	assert.Equal(t, vmcommon.UserError, retCode)
+}
+
+func TestStakingSC_ExecuteStake(t *testing.T) {
+	t.Parallel()
+
+	expectedRegistrationData := stakingData{
+		StartNonce:    100,
+		Staked:        true,
+		UnStakedNonce: 0,
+		BlsPubKey:     []byte{100},
+		StakeValue:    big.NewInt(0),
+	}
+
+	stakeValue := big.NewInt(100)
+	eei, _ := NewVMContext(&mock.BlockChainHookStub{}, &mock.CryptoHookStub{})
+	eei.SetSCAddress([]byte("addr"))
+
+	stakingSmartContract, _ := NewStakingSmartContract(stakeValue, eei)
+	arguments := CreateVmContractCallInput()
+	arguments.Function = "stake"
+	arguments.Arguments = []*big.Int{big.NewInt(100), big.NewInt(200)}
+	arguments.Header = &vmcommon.SCCallHeader{Number: big.NewInt(100)}
+	arguments.CallValue = big.NewInt(100)
+
+	retCode := stakingSmartContract.Execute(arguments)
+	assert.Equal(t, vmcommon.Ok, retCode)
+
+	var registrationData stakingData
+	data := stakingSmartContract.eei.GetStorage(arguments.CallerAddr)
+	err := json.Unmarshal(data, &registrationData)
+	assert.Nil(t, err)
+	assert.Equal(t, expectedRegistrationData, registrationData)
+
+	receiverBalance := stakingSmartContract.eei.GetBalance(arguments.RecipientAddr)
+	senderBalance := stakingSmartContract.eei.GetBalance(arguments.CallerAddr)
+	assert.Equal(t, big.NewInt(-100), senderBalance)
```

Comment:
why -100 is not ok because when you transfer balance if account not exist will be created an account with 0 balance and from 0 is subtract 100 and sender balance will be -100 ?

Reply:
-100 is ok. As these are local changes. Not actual account balance values.

Reply:
there should never exist negative balances. If I stake from an account that does not exist, should the stakeing work?

Reply:
But `eei.GetBalance()` returns the real balance of an account. According to the local Node, indeed, but it's still a real balance in ERD, right?